### ISSUE TITLE
Simplify TravisCI

### ISSUE THREAD

**Issue by edgarriba** [tiny-dnn/tiny-dnn#984]
Need to simplify continuous integration test since sometimes it just crashes.

- Suggest to use docker for development
https://hub.docker.com/r/tinydnn/tinydnn/
- Explore staged building
https://docs.travis-ci.com/user/build-stages/share-docker-image
- Remove cppcheck: what's the alternative to end up with unstructured code ? @beru may have some opinions on this.

# NOTE: move this discussion to: https://groups.google.com/forum/#!forum/tiny-dnn-dev/new

**Comment by beru** [comment]
@edgarriba 

> Remove cppcheck: what's the alternative to end up with unstructured code ? @beru may have some opinions on this.

Precisely, I didn't mention about cppcheck. What I had in mind was about `clang-format` check (because most of the people who send pull requests simply don't know about it and same problem happened again and again... ).

After all, since you kindly commented out [these lines](https://github.com/tiny-dnn/tiny-dnn/commit/c5afeda82f4c09d395257761eb118aec11194b25#diff-354f30a63fb0907d4ad57269548329e3L99), `scripts/run_lints.sh` file won't be called anymore so both of `cpplint` and `clang-format` checks are disabled now and I don't have an objection against it.

**Comment by beru** [comment]
@edgarriba 
> Remove cppcheck: what's the alternative to end up with unstructured code ? @beru may have some opinions on this.

To this question, I've already answered at [here](https://github.com/tiny-dnn/tiny-dnn/pull/978#issuecomment-414293064). And I think you responded it by commiting c5afeda82f4c09d395257761eb118aec11194b25.

>> we can remove it. But we what's the alternative to not get a messy and unstructured code ?

> Maintainers like us occasionally may apply scripts/clang-format.sh script.
My opinion is new contributors shouldn't be bothered with it.

When new people post pull requests for enhancements, they naturally don't know a lot of things that maintainers added. Adding rules or libraries might sound attractive at first but when people fail to follow them, they're nothing but burdens.

**Comment by bhack** [comment]
Can we use a precommit hook like in [GNOME](https://gitlab.gnome.org/GNOME/gjs/commit/dbd555e865e73ee37bff077328328f920dc6f98c)?

**Comment by beru** [comment]
@bhack I don't think that's a great idea.

- Not everyone have specific version of `clang-format` that we use on their machines.
- Maintaining the scripts for several operating systems would be needed.
- People will have difficulties when they try to commit even to their local git repositories.

Do we really need to have consistent coding styles with **every commits** in master branch?

What are the actual benefits of forcing/keeping same coding styles? Is it that important?

I gues it's too cumbersome for people who post pull requests. New people don't even know the rule.

**Comment by beru** [comment]
@edgarriba @bhack I repeatedly wrote my opinions about this problems several times in the past.

Please think twice before adding rules.

**Comment by bhack** [comment]
@beru I understand you point of view but also people don't like to reformat the code that they pull from master especially if they have the linter integrater with their dev enivronment/IDE. Also a developer that reformat the code pull from master to open a PR I thing that will create a noise in the PR if the reformat impact is quite high.
I don't know if it needs to be mandatory to submit tests but it is not just Gnome. Just to evaluate other two examples:

https://github.com/pytorch/glow/blob/master/docs/CodingStandards.md#format-your-code
https://github.com/tensorflow/tensorflow/blob/master/CONTRIBUTING.md#c-coding-style

**Comment by beru** [comment]
> I understand you point of view but also people don't like to reformat the code that they pull from master especially if they have the linter integrater with their dev enivronment/IDE. 

Whey do they need to reformat pulled tiny-dnn code? We don't reformat third-party library files.
And I don't think other library maintainers use same `.clang-format` file.

Does their environment/IDE automatically reformat tiny-dnn files because of `.clang-format` file in the root folder? Does it often happen? I guess they should change the IDE/project configuration in that case...

> Also a developer that reformat the code pull from master to open a PR I thing that will create a noise in the PR if the reformat impact is quite high.

In that case, maitainers should update master branch and ask PR submitters to rebase PR branch to the master branch. I hope it won't frequently happen.

> I don't know if it needs to be mandatory to submit tests but it is not just Gnome. Just to evaluate other two examples:

I think writing tests should be recommended, not compulsory.

From pytorch document you introduced,  https://github.com/pytorch/glow/blob/master/docs/CodingStandards.md#format-your-code

> We **encourage** the use of clang-format to enforce code style and formatting. Commits that only change the formatting of code should go in separate commits. This makes reviewing the code and inspecting history easier.

> It's **recommended** to use a pre-commit hook to properly format your code prior committing any changes. Run the following command from the root of the repo to enable the hook: ln -s ../../utils/format.sh .git/hooks/pre-commit.

> https://github.com/tensorflow/tensorflow/blob/master/CONTRIBUTING.md#c-coding-style

It seems TensorFlow team has wealthy guidelines.
